### PR TITLE
[macOS] Add run release test in devicelab

### PR DESCRIPTION
**Pull request by cbracken**
Adds a test that invokes flutter run in release mode on macOS desktop,
waits for successful launch and the flutter command list, then sends the
'q' command to quit the running app.

This adds an integration test for https://github.com/flutter/flutter/pull/100504.

Issue: https://github.com/flutter/flutter/issues/100348 (fix)
Issue: https://github.com/flutter/flutter/issues/97978 (partial fix)
Issue: https://github.com/flutter/flutter/issues/97977 (partial fix)
Umbrella issue: https://github.com/flutter/flutter/issues/60113

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by cbracken** [comment]
Requesting a test exemption from @Hixie until the test exemption bot is updated. 

Exemption reason: This is a test :)

File: dev/devicelab/bin/tasks/run_release_test_macos.dart [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+import 'package:flutter_devicelab/framework/devices.dart';
+import 'package:flutter_devicelab/framework/framework.dart';
+import 'package:flutter_devicelab/framework/task_result.dart';
+import 'package:flutter_devicelab/framework/utils.dart';
+import 'package:path/path.dart' as path;
+
+/// Basic launch test for desktop operating systems.
+void main() {
+  task(() async {
+    deviceOperatingSystem = DeviceOperatingSystem.macos;
```
Comment (jmagman) [inline dev/devicelab/bin/tasks/run_release_test_macos.dart:19]: You don't have to do it in this PR but can you also replace this `deviceIdOverride` hack with your nicer `DeviceOperatingSystem.macos` fix? 
https://github.com/flutter/flutter/blob/deacb8d434b4b05291be912da2c4bcac869237cf/dev/devicelab/bin/tasks/hot_mode_dev_cycle_macos_target__benchmark.dart#L9

File: dev/devicelab/bin/tasks/run_release_test_macos.dart [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    final Device device = await devices.workingDevice;
+    final Directory appDir = dir(path.join(flutterDirectory.path, 'dev/integration_tests/ui'));
+    await inDirectory(appDir, () async {
+      final Completer<void> ready = Completer<void>();
+      final List<String> stdout = <String>[];
+      final List<String> stderr = <String>[];
+
+      print('run: starting...');
+      final Process run = await startProcess(
```
Comment (jmagman) [inline dev/devicelab/bin/tasks/run_release_test_macos.dart:28]: Use `startFlutter` instead, it handles `--local-engine`.  You may need update it to pipe `isBot` through that though.
https://github.com/flutter/flutter/blob/9072a09958a3fd9eaf4a857acbcc4b4e72167bbd/dev/devicelab/lib/tasks/microbenchmarks.dart#L39

**Review by jmagman — COMMENTED** [review]
```
warning â�¢ Missing case clause for 'macos' â�¢ dev/devicelab/lib/tasks/perf_tests.dart:585:7 â�¢ missing_enum_constant_in_switch
warning â�¢ Missing case clause for 'macos' â�¢ dev/devicelab/lib/tasks/perf_tests.dart:705:7 â�¢ missing_enum_constant_in_switch
warning â�¢ Missing case clause for 'macos' â�¢ dev/devicelab/lib/tasks/perf_tests.dart:1299:5 â�¢ missing_enum_constant_in_switch
  error â�¢ The non-nullable local variable 'releaseSizeInBytes' must be assigned before it can be used â�¢ dev/devicelab/lib/tasks/perf_tests.dart:1363:29 â�¢ not_assigned_potentially_non_nullable_local_variable
warning â�¢ Missing case clause for 'macos' â�¢ dev/devicelab/lib/tasks/perf_tests.dart:1378:5 â�¢ missing_enum_constant_in_switch
```

File: TESTOWNERS [hunk 3]
```diff
@@ -123,6 +123,7 @@
 /dev/devicelab/bin/tasks/platform_interaction_test.dart @stuartmorgan @flutter/plugin
 /dev/devicelab/bin/tasks/platform_view__start_up.dart @zanderso @flutter/engine
 /dev/devicelab/bin/tasks/run_release_test.dart @zanderso @flutter/tool
+/dev/devicelab/bin/tasks/run_release_test_macos.dart @cbracken @flutter/tool
```
Comment (keyonghan) [inline TESTOWNERS:126]: Nit: put under section `## Host only DeviceLab tests`

**Review by keyonghan — APPROVED** [review]
Config LGTM with nits.

File: dev/devicelab/bin/tasks/run_release_test_macos.dart (hunk 2 above)
Comment (cbracken) [reply dev/devicelab/bin/tasks/run_release_test_macos.dart]: Done.

File: dev/devicelab/bin/tasks/run_release_test_macos.dart (hunk 1 above)
Comment (cbracken) [reply dev/devicelab/bin/tasks/run_release_test_macos.dart:19]: Thanks for spotting this. Looks like we do the same in the Windows/Linux equivalents too. I'll send one patch that does all three and adds the Linux DeviceOperatingSystem if it doesn't exist (the Windows one definitely does).

File: .ci.yaml [hunk 4]
```diff
@@ -3684,6 +3684,26 @@ targets:
...
+        ]
+      tags: >
+        ["devicelab","hostonly"]
+      task_name: run_release_test_macos
+    scheduler: luci
+    runIf:
+    - dev/**
+    - packages/flutter_tools/**
+    - bin/**
+    - .ci.yaml
```
[the remark this answers is not in the document]
Comment (cbracken) [reply .ci.yaml]: Done. Also done for the test above, which also had bad indentation.

File: TESTOWNERS (hunk 3 above)
Comment (cbracken) [reply TESTOWNERS]: Done. Thanks!

**Comment by Hixie** [comment]
test-exempt: is a test

File: dev/devicelab/bin/tasks/run_release_test_macos.dart [hunk 5]
```diff
@@ -25,10 +25,14 @@ void main() {
       final List<String> stderr = <String>[];
 
       print('run: starting...');
-      final Process run = await startProcess(
-        path.join(flutterDirectory.path, 'bin', 'flutter'),
-        <String>['--suppress-analytics', 'run', '--release', '-d', device.deviceId, 'lib/main.dart'],
-        isBot: false, // we just want to test the output, not have any debugging info
+      final List<String> options = <String>[
```
[the remark this answers is not in the document]
Comment (cbracken) [reply dev/devicelab/bin/tasks/run_release_test_macos.dart:28]: `startFlutter` doesn't appear to have an `isBot` parameter, but yes, it passes without it; I could add it if needed.
```
[run_release_test_macos] Process terminated with exit code 0.
Task result:
{
  "success": true,
  "data": null,
  "detailFiles": [],
  "benchmarkScoreKeys": [],
  "reason": "success"
}


═══════════════╡ ••• Finished task "run_release_test_macos" ••• ╞═══════════════
```

File: dev/devicelab/bin/tasks/run_release_test_macos.dart [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+import 'package:flutter_devicelab/framework/task_result.dart';
+import 'package:flutter_devicelab/framework/utils.dart';
+import 'package:path/path.dart' as path;
+
+/// Basic launch test for desktop operating systems.
+void main() {
+  task(() async {
+    deviceOperatingSystem = DeviceOperatingSystem.macos;
+    final Device device = await devices.workingDevice;
+    final Directory appDir = dir(path.join(flutterDirectory.path, 'dev/integration_tests/ui'));
```
[the remark this answers is not in the document]
Comment (cbracken) [reply dev/devicelab/bin/tasks/run_release_test_macos.dart:21]: Done.

**Review by jmagman — APPROVED** [review]
LGTM, thanks!

**Comment by jmagman** [comment]
This is failing:
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8818956881147077521/+/u/run_run_release_test_macos/test_stdout

macOS builds spew so much to stderr...

**Comment by cbracken** [comment]
Workaround for the failure landed in https://github.com/flutter/flutter/pull/100591.

**Pull request merged**